### PR TITLE
Provide an RSocketMessageHandlerCustomizer to allow customizing of the RSocketMessageHandler

### PR DESCRIPTION
**Pull request by aartiguptaa**
Provide an RSocketMessageHandlerCustomizer to allow customizing of the RSocketMessageHandler
#20303

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@aartiguptaa Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=21081)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-boot&pullRequestId=21081) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@aartiguptaa Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=21081)!

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketMessageHandlerCustomizer.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.util.RouteMatcher;
+
+/**
+ * @author Aarti Gupta
+ * Callback interface that can be used to customize a RSocketMessageHandler {@link Connector}.
+ */
+@FunctionalInterface
+public interface RSocketMessageHandlerCustomizer {
+
+	RSocketMessageHandler setRouteMatcher(RouteMatcher handler);
```
Comment (mbhave) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketMessageHandlerCustomizer.java:16]: To be consistent with other customizers, the method shouldn't return anything but be `void` instead.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketMessagingAutoConfiguration.java [hunk 2]
```diff
@@ -36,17 +40,19 @@
  * @author Brian Clozel
  * @since 2.2.0
  */
-@Configuration(proxyBeanMethods = false)
-@ConditionalOnClass({ RSocketRequester.class, RSocketFactory.class, TcpServerTransport.class })
+@Configuration(proxyBeanMethods = true)
```
Comment (mbhave) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketMessagingAutoConfiguration.java:43]: Is there a reason this was change to `true`?

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketMessagingAutoConfiguration.java [hunk 3]
```diff
@@ -36,17 +40,19 @@
...
 	@Bean
-	@ConditionalOnMissingBean
-	public RSocketMessageHandler messageHandler(RSocketStrategies rSocketStrategies) {
-		RSocketMessageHandler messageHandler = new RSocketMessageHandler();
-		messageHandler.setRSocketStrategies(rSocketStrategies);
-		return messageHandler;
+	@ConditionalOnMissingBean(RSocketMessageHandler.class)
+	public RSocketMessageHandler messageHandler(RSocketStrategies rSocketStrategies, ObjectProvider<RSocketMessageHandlerCustomizer> customizers) {
+		RSocketMessageHandlerCustomizer rSocketMessageHandlerCustomizer = customizers.getIfAvailable();
+		return rSocketMessageHandlerCustomizer.setRouteMatcher(rSocketStrategies.routeMatcher());
```
Comment (mbhave) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketMessagingAutoConfiguration.java:53]: It looks like `messageHandler.setRSocketStrategies(rSocketStrategies)` got removed as part of this change. We still want to do that and apply the customizers after that.

**Comment by mbhave** [comment]
@aartiguptaa Thanks for the pull request. I've left a few comments on it. In addition to those, the pull request is missing tests to verify that the customizers are being applied to the `RSocketMessageHandler`. Tests for that would need to be added to [this class](https://github.com/spring-projects/spring-boot/blob/7d7b1e13a2409a9a5499c87fcfe73c757b0cea09/spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/rsocket/RSocketMessagingAutoConfigurationTests.java#L37). They would look similar to the tests that verify that other customizers have been applied. See [this example](https://github.com/spring-projects/spring-boot/blob/3e6c15c451ab39ab51c6d96b4110b2e5fd2bbd19/spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/web/reactive/WebFluxAutoConfigurationTests.java#L124).

**Comment by mbhave** [comment]
@aartiguptaa Thank you for making your first contribution to Spring Boot. It has now been merged into master along with these [amendments](https://github.com/spring-projects/spring-boot/commit/6007a71b9cb37cd2e2cd5b1e2e35e2ecfdb0ba1b).

**Pull request closed without merging**